Update markdown reference (#93)
codewars/docs: content/references/markdown/index.md

next: /references/markdown/extensions/
---

# Basic Markdown Syntax

Codewars supports [Markdown][wiki-markdown].

More specifically, [GitHub Flavored Markdown][gfm] (strict superset of [CommonMark][common-mark]) and few [Codewars extensions][extensions].

You may find below a summary of the most used ones.

## Emphasis

> ```code

Reviewer: ```suggestion
```markdown
```

_This is italic_
*This is also italic*

__This is bold__
**This is bold too**

___here, italic and bold___
***here, italic and bold***
```
Note that `*You **can** combine them*`: *You **can** combine them*.

## Headers

> ```code

Reviewer: ```suggestion
```markdown
```

# This is &lt;h1&gt;
## This is &lt;h2&gt;
###### This is &lt;h6&gt;
```

## Code blocks

To get text displayed without the renderer playing with the indentation or whitespaces, code blocks

### Inline code blocks

To highlight code  ` \`inline code block\` `

## Lists

### Unordered lists:

> ```code

Reviewer: ```suggestion
```markdown
```

* item 1
* item 2
    * subitem 2.1
    * subitem 2.2
```

You can use different marks to get same result, like `+` or `-`.

### Ordered lists:

> The numbering is automatic, hence you cannot mix different types of paragraphs or items if you want to keep the numbering correct.

Reviewer: I don't think we should include this for this basic syntax cheat sheet page.

Author: removed

> ```code

Reviewer: ```suggestion
```markdown
```

1. item 1
1. item 2
  * subitem 2.1
  * subitem 2.2
```

Example:

1. item 1
1. item 2
    1. item 2.1
    * unordered mark
    1. cannot be item 2.2

## Links

Valid url links are automatically converted to links. If you want to have a custom text instead of the url, you can you this syntax:

```code
[Custom text](address)

> ```

Reviewer: ~~~suggestion
```markdown
[link text](https://www.example.com)
```
~~~

Like this [custom text leading to Codewars](https://www.codewars.com/dashboard).

## Images

Similarly to links:

> ```code

Reviewer: ```suggestion
```markdown
```

![alt text](address)
```

## Paragraphs & line continuations:

* To get separated paragraphs, put an empty line between your paragraphs in the text.
    If you just use a line break, both parapgraphs will be merged together when rendered, like those two lines even if they are written on two different lines.

* If you just want a line break without any space between the paragraphs, put two whitespaces  
    ...at the end of the line, like it is here.

The related code is:

```code
* To get separated [...] in the text.
    If you just use [...] different lines.

* If you just want a line break without any space between the paragraphs, put two whitespaces  
    ...at the end of the line, like it is here.
```

## Blockquotes

To quote the answer of another user, for example:

> ```code

Reviewer: ```suggestion
```markdown
```

> > You quote like this...

Reviewer: ```suggestion
> Quoted text
```

```

> > ...and you get this.

Reviewer: ```suggestion
> Quoted text
```

next: /references/markdown/extensions/
---

# Basic Markdown Syntax

Codewars supports [Markdown][wiki-markdown].

More specifically, [GitHub Flavored Markdown][gfm] (strict superset of [CommonMark][common-mark]) and few [Codewars extensions][extensions].

You may find below a summary of the most used ones.

## Emphasis

```markdown
_This is italic_
*This is also italic*

__This is bold__
**This is bold too**

___here, italic and bold___
***here, italic and bold***

> ```
> Note that `*You **can** combine them*`: *You **can** combine them*.

Reviewer: ````suggestion
*You can have **bold** in italic*
```
````

## Displaying code

Use code spans and code blocks to show the text as is. Use code block to show the syntax highlighted code.

### Code span

Use single backtick (\`) to show the text as is inline.

```markdown
Code span example: `print("example")`
```

### Code block

For multiline code blocks, use triple backticks.

````markdown
```
This is a code block
       you can indent inside it as you like.
```
````

> Info strings can be added on the opening triple backtick, to get the syntax highlighted. The info string must be a valid language name (`pyhton`, `javascript`, `php`, `ruby`, `cpp`, ...).

Reviewer: ```suggestion
The opening code fence can be followed by an optional text called the info string. When the info string is a valid language ID, `python` for example, the content is syntax highlighted.
See the language's documentation in [supported languages](/languages/) to find the langauge ID.
```

````markdown
```python
def hello_world():
    # this is rendered
    hi = lambda: 'hello'
    print(hi())
    return 42 
```
````

> See Codewars' [extensions] page about sequential and language specific blocks.

Reviewer: ```suggestion
If you're writing a kata description, see [Codewars' extensions][extensions] like sequential and language specific blocks.
```

## Lists

### Unordered lists

```markdown
* item 1
* item 2
    * subitem 2.1
    * subitem 2.2
```

> You can use different marks to get same result, like `+` or `-`.

Reviewer: ```suggestion
You can use `-`, `*`, and `+` as bullet markers.
```

### Ordered lists

```markdown

> 1. item 1
> 1. item 2
>   * subitem 2.1
>   * subitem 2.2

Reviewer: ```suggestion
1. item 1
    1. subitem 1.1
2. item 2
    * subitem 2.1

or

1. item 1
    1. subitem 1.1
1. item 2 (incrementing the marker number is optional)
    * subitem 2.1
```

```

> 
> Example:
> 
> 1. item 1
> 1. item 2
>     1. item 2.1
>     * unordered mark
>     1. cannot be item 2.2
> 

Reviewer: ```suggestion
```

## Links

Valid URLs are automatically converted to links with link text set to the URL.

To specify a link text, use the following syntax:

```markdown
[link text](https://www.example.com)
```

> Like this [custom text leading to Codewars](https://www.codewars.com/dashboard).

Reviewer: ```suggestion
```
Inconsistent with other sections.

## Images

> Similarly to links:

Reviewer: ```suggestion
To include an image:
```

```markdown
![alt text](address)
```

## Blockquotes

To quote the answer of another user, for example:

```markdown
> Quoted text
```

> > Quoted text

Reviewer: ```suggestion
```
Inconsistent with other sections.

<!--

> TODO Finish this basic Markdown reference by listing most frequently used ones

Reviewer: ```suggestion
```

next: /references/markdown/extensions/
---

# Basic Markdown Syntax

Codewars supports [Markdown][wiki-markdown].

More specifically, [GitHub Flavored Markdown][gfm] (strict superset of [CommonMark][common-mark]) and few [Codewars extensions][extensions].

You may find below a summary of the most used ones.

## Emphasis

```markdown
_This is italic_
*This is also italic*

__This is bold__
**This is bold too**

___here, italic and bold___
***here, italic and bold***
```
Note that `*You **can** combine them*`: *You **can** combine them*.

## Headers

```markdown
# This is &lt;h1&gt;
## This is &lt;h2&gt;
###### This is &lt;h6&gt;

> ```

Reviewer: ~~~suggestion
```text
# This is <h1>
## This is <h2>
###### This is <h3>
```
~~~

I think this will work.

## Code blocks

To get text displayed without the renderer playing with the indentation or whitespaces, code blocks

### Inline code blocks:

To highlight code inside a sentence, use simple backticks  ``` `print("This is some code")` ```.

> ### Mutliline code blocks:

Reviewer: ~~~suggestion
## Displaying code

Use code spans and code blocks to show the text as is. Use code block to show the syntax highlighted code.

### Code span

Use single backtick (<code>`</code>) to show the text as is inline.

```markdown
Code span example: `print("example")`
```

### Code block
~~~

For multiline code blocks, use triple backticks.

Constraints:

>    - Put an empty line before and after the code block itself, so that it's rendered correctly

Reviewer: Is this Codewars' bug? I don't think empty lines are necessary.

Author: The only thing I know is that I'm almost always lost about how to do this, between gitter, CW and here... :/

But currently, on codewars, it's definitely required otherwise youget some mess like that: https://www.codewars.com/kata/56941f177c0a52aef50000a2

Reviewer: That's because of all the mess it has:

![image](https://user-images.githubusercontent.com/639336/88430456-522df000-cdad-11ea-9f59-a16e8ecd700b.png)

Reviewer: I'll update it to check if it's really a bug.

Author: mmmh... I'm not a trustful source apparently... ;o

Author: not about this kata: voile messed it up (uneditable through the edit panel) by playing with the end points to change it's rank. See here: https://github.com/codewars/codewars.com/issues/1939#issuecomment-536228782

Reviewer: With proper Markdown https://www.codewars.com/kata/56941f177c0a52aef50000a2

Reviewer: Ugh, I didn't see your comment and didn't realize it had that issue. Now it's in beta.

Reviewer: OK, should be fixed now. I couldn't re-approve normally because it had open issues.

next: /references/markdown/extensions/
---

# Basic Markdown Syntax

Codewars supports [Markdown][wiki-markdown].

More specifically, [GitHub Flavored Markdown][gfm] (strict superset of [CommonMark][common-mark]) and few [Codewars extensions][extensions].

You may find below a summary of the most used ones.

## Emphasis

```markdown
_This is italic_
*This is also italic*

__This is bold__
**This is bold too**

___here, italic and bold___
***here, italic and bold***
```
Note that `*You **can** combine them*`: *You **can** combine them*.

> ## Headers
> 
> <pre><code>
> # This is &lt;h1&gt;
> ## This is &lt;h2&gt;
> ### This is &lt;h3&gt;
> </code></pre>
> 

Reviewer: ```suggestion
```

Let's just remove headers. It's not that useful for writing comments and it's styled like bold text in comments so it's misleading.

We can have a page for writing kata descriptions, and we can mention headers there.

## Displaying code

Use code spans and code blocks to show the text as is. Use code block to show the syntax highlighted code.

### Code span

> Use single backtick `` ` `` to show the text as is inline.

Reviewer: ```suggestion
Use single backtick (\`) to show the text as is inline.
```

Reviewer: This is because the current code span is style with surrounding backticks. Using code span will show three backticks.

next: /references/markdown/extensions/
---

# Basic Markdown Syntax

Codewars supports [Markdown][wiki-markdown].

More specifically, [GitHub Flavored Markdown][gfm] (strict superset of [CommonMark][common-mark]) and few [Codewars extensions][extensions].

You may find below a summary of the most used ones.

## Emphasis

```markdown
_This is italic_
*This is also italic*

__This is bold__
**This is bold too**

___here, italic and bold___
***here, italic and bold***
```
Note that `*You **can** combine them*`: *You **can** combine them*.

## Headers

> ```text
> # This is <h1>
> ## This is <h2>
> ### This is <h3>
> ```

Reviewer: ```suggestion
<pre><code>
# This is &lt;h1&gt;
## This is &lt;h2&gt;
### This is &lt;h3&gt;
</code></pre>
```

Author: damn it... x/

## Displaying code

Use code spans and code blocks to show the text as is. Use code block to show the syntax highlighted code.

### Code span

> Use single backtick (`) to show the text as is inline.

Reviewer: ```suggestion
Use single backtick (\`) to show the text as is inline.
```

> Use single backtick (<code>`</code>) to show the text as is inline.

Reviewer: ```suggestion
Use single backtick (`) to show the text as is inline.
```

```markdown
Code span example: `print("example")`
```

### Code block

For multiline code blocks, use triple backticks.

> Constraints:
>    - Put an empty line before and after the code block itself, so that it's rendered correctly
>    - The triple backticks must be alone one their line (see [extensions](./references/markdown/extensions/) for exceptions)

Reviewer: ```suggestion
```

Author: sure you wanna remove those? Because they are pars of the tiny tricks to write descriptions :o

Reviewer: > Because they are part of the tiny tricks to write descriptions :o

I'm not sure what you mean. Code blocks should work without empty lines.

I'm not sure what you mean on the second one. Can you explain? Maybe I'm forgetting something so examples of what works and what doesn't will help.

Author: wth x/ I lost my answer... x// Here it goes again:

> tricks

just things that one encounters from time to time when writing comments or descriptions. Unexpected behaviors that are a hassle to avoid, like the "non-" or "mis-rendering" of code blocks, bullet points acting ([see example here](https://www.codewars.com/kata/5efdde81543b3b00153e918a/discuss#5f1bcdbfbe281b002a2e7490); code used for the related message at the end of the present message), and stuff like that. Often, I do't even know the real cause behind those troubles.

> Can you explain the second one?

See examples in the same link as above.

---

````markdown
___Ignore this message, it's a test___

---

# about bullet points

* I'd change the formatting here, grouping/adding several subpoints: To encrypt the characters, you'll have to build pairs of characters. To do so:
  * merge all the characters
  * talk about the x stuff there.
  * "add x at the end if needed"
  * then break into pairs
* formatting: for 1) and 2), add extra whitespace so that the indentation is correct
* I'd add at the end of the previous point the note about the consecutive Xs: "Don't worry about inputs with consecutive xx, you'll never encounter them during the tests". Could be enough like that.
* the last point is curre ntly ununderstandable to me x) Maybe.... "Now you insert space characters in the ciphered string so that they are matching there original position in the input string."? (this way, that makes clear what to do about the remaining ciphered letters at the same time)
* (you will never encounter 'x' in middle of message for the sake of simplicity) To remove, since talked about before.

* check you have a fixed test with a plain text like abc cdem, so that the mix of requirements about whitespaces and addint the x between consecutive letters is made clearer/tested for sure

---

* I'd change the formatting here, grouping/adding several subpoints: To encrypt the characters, you'll have to build pairs of characters. To do so:
  * merge all the characters
  * talk about the x stuff there.
  * "add x at the end if needed"
  * then break into pairs
* formatting: for 1) and 2), add extra whitespace so that the indentation is correct
* I'd add at the end of the previous point the note about the consecutive Xs: "Don't worry about inputs with consecutive xx, you'll never encounter them during the tests". Could be enough like that.
* the last point is curre ntly ununderstandable to me x) Maybe.... "Now you insert space characters in the ciphered string so that they are matching there original position in the input string."? (this way, that makes clear what to do about the remaining ciphered letters at the same time)
* (you will never encounter 'x' in middle of message for the sake of simplicity) To remove, since talked about before.

check you have a fixed test with a plain text like abc cdem, so that the mix of requirements about whitespaces and addint the x between consecutive letters is made clearer/tested for sure

---

# about code blocks

works:
```
def pythonFunc():
    return 0
```

```
def pythonFunc():
    return 0```    <<< here

wrong ending of the block
```

some text here

first line is taken for a language specific block and is swallowed:

```def pythonFunc():
    return 0
```

````

Reviewer: For the bullet points rendered weirdly:
![image](https://user-images.githubusercontent.com/639336/88588522-8c48fd00-d00c-11ea-9f53-34e4c8eb5d38.png)

It's a bug in CW's CSS. It't tricky to style lists from Markdown because it can add `<p>` inside `<li>`. I'm going to fix it by using `@tailwincss/typography` plugin like this docs sometime soon.

---

For code block, I didn't understand what you meant, but the example helped.

I didn't understand because I never thought of writing that way, knowing the Markdown spec and the example seemed clear enough. I also got confused because you wrote that extensions are exceptions (it's not).

The opening triple backticks are followed by _optional_ infostring. The infostrings are usually language ids, but it can be anything.

````markdown
```def python_func():
  return 0
```
````
So the above is a code block containing just `  return 0` with `infostring` `def python_func():`.

````markdown
```
def python_func():
    return 0```
````
is invalid because the closing backticks must be on its own line.

I think we should have an example with infostring and also mention that it should be valid language id in order to get it syntax highlighted.

Reviewer: I fixed the CSS bug for Markdown and it's currently on preview. Your lists example is not broken anymore: https://preview.codewars.com/kata/5efdde81543b3b00153e918a/discuss#5f1bcdbfbe281b002a2e7490

I had to change more than I wanted, so I hope I didn't break anything else.

Author: > I think we should have an example with infostring and also mention that it should be valid language id in order to get it syntax highlighted.

:+1: I'll take care of that later.

(I actually faced that kind of situation with newcomers who couldn't format their code correctly in the comments)

````markdown
```
This is a code block
       you can indent inside it as you like.
```
````

> Another way to declare a code block is to put 4 whitesapces at the beginning of the line
> 
> ````markdown
>    This will be rendered as a code block too
>            you can indent inside it as you like.
> 
> ````

Reviewer: ```suggestion
```

This is true, but I don't think we need it on this page. Fenced code blocks will be enough.

Author: ok. I'll commit after you asnwer to the previous one

## Lists

> ### Unordered lists:

Reviewer: ```suggestion
### Unordered lists
```

```markdown
* item 1
* item 2
    * subitem 2.1
    * subitem 2.2
```

You can use different marks to get same result, like `+` or `-`.

> ### Ordered lists:

Reviewer: ```suggestion
### Ordered lists
```

```markdown
1. item 1
1. item 2
  * subitem 2.1
  * subitem 2.2
```

Example:

1. item 1
1. item 2
    1. item 2.1
    * unordered mark
    1. cannot be item 2.2

## Links

> Valid url links are automatically converted to links. If you want to have a custom text instead of the url, you can you this syntax:

Reviewer: ```suggestion
Valid URLs are automatically converted to links with link text set to the URL.

To specify a link text, use the following syntax:
```

```markdown
[link text](https://www.example.com)
```

Like this [custom text leading to Codewars](https://www.codewars.com/dashboard).

## Images

Similarly to links:

```markdown
![alt text](address)
```

> ## Paragraphs & line continuations:

Reviewer: This section is too confusing.
Basically, to start a new paragraph you need to have an empty line. I think the two spaces to introduce line break is too much for this page.

Author: Yeah, I had trouble just to write somehting about that... :/ I guess we just should get rid of that part.
